### ISSUE TITLE
Some invalid inputs cause StackOverflowError in ballerina parser

### ISSUE THREAD

**Issue by kdsuneraavinash** [ballerina-platform/ballerina-lang#27057]
**Description:**

Some invalid inputs will cause Stackoverflow Exceptions in the ballerina parser. 
Formatters and other tools depending on the parser crash because of this.
(eg: VSCode extension formatter will freeze)

**Steps to reproduce:**

Run the following code segment using the `ballerina run` command.

```ballerina
public function main() {
    return enum Color{A, B, C};
}
```

After about 30-40 seconds, a bad sad error will be throw. Following is the error log.

```text
[2020-11-22 01:55:07,480] SEVERE {b7a.log.crash} - null 
java.lang.StackOverflowError
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExprRelatedAlternativePaths(BallerinaParserErrorHandler.java:2044)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInStmtRelatedAlternativePaths(BallerinaParserErrorHandler.java:1904)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1668)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1131)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExprRelatedAlternativePaths(BallerinaParserErrorHandler.java:2044)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInStmtRelatedAlternativePaths(BallerinaParserErrorHandler.java:1904)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1668)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1131)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExpressionRhs(BallerinaParserErrorHandler.java:2274)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExprRelatedAlternativePaths(BallerinaParserErrorHandler.java:2037)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInStmtRelatedAlternativePaths(BallerinaParserErrorHandler.java:1904)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1668)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1131)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:387)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:328)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1142)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExprRelatedAlternativePaths(BallerinaParserErrorHandler.java:2044)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInStmtRelatedAlternativePaths(BallerinaParserErrorHandler.java:1904)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1668)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1131)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExprRelatedAlternativePaths(BallerinaParserErrorHandler.java:2044)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInStmtRelatedAlternativePaths(BallerinaParserErrorHandler.java:1904)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1668)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1131)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExpressionRhs(BallerinaParserErrorHandler.java:2274)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExprRelatedAlternativePaths(BallerinaParserErrorHandler.java:2037)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInStmtRelatedAlternativePaths(BallerinaParserErrorHandler.java:1904)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1668)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1131)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:387)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:328)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1142)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExprRelatedAlternativePaths(BallerinaParserErrorHandler.java:2044)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInStmtRelatedAlternativePaths(BallerinaParserErrorHandler.java:1904)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1668)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1131)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExprRelatedAlternativePaths(BallerinaParserErrorHandler.java:2044)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInStmtRelatedAlternativePaths(BallerinaParserErrorHandler.java:1904)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1668)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1131)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExpressionRhs(BallerinaParserErrorHandler.java:2274)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExprRelatedAlternativePaths(BallerinaParserErrorHandler.java:2037)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInStmtRelatedAlternativePaths(BallerinaParserErrorHandler.java:1904)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1668)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1131)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:387)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:328)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1142)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExprRelatedAlternativePaths(BallerinaParserErrorHandler.java:2044)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInStmtRelatedAlternativePaths(BallerinaParserErrorHandler.java:1904)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1668)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1131)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExprRelatedAlternativePaths(BallerinaParserErrorHandler.java:2044)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInStmtRelatedAlternativePaths(BallerinaParserErrorHandler.java:1904)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1668)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1131)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExpressionRhs(BallerinaParserErrorHandler.java:2274)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInExprRelatedAlternativePaths(BallerinaParserErrorHandler.java:2037)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInStmtRelatedAlternativePaths(BallerinaParserErrorHandler.java:1904)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1668)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1131)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:387)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:328)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1142)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1672)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1131)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatch(AbstractParserErrorHandler.java:181)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.recover(AbstractParserErrorHandler.java:102)
	at io.ballerina.compiler.internal.parser.AbstractParser.recover(AbstractParser.java:106)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseFunctionQualifiers(BallerinaParser.java:5070)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseFunctionQualifiers(BallerinaParser.java:5071)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseFunctionQualifiers(BallerinaParser.java:5071)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseFunctionQualifiers(BallerinaParser.java:5071)
        --- the same line repeated for about 1000 times ---
```

Following is another code segment with this issue. (Should be the only statement in the file and no semicolon should be present.

```ballerina
panic error("Record is nil")
```

**Affected Versions:**

Present in both the `Ballerina Swan Lake Preview 5` and the master branch.

**OS, DB, other environment details and versions:**

Checked in Manjaro (kernel 5.4.74) 
Java version:
```
openjdk 11.0.8 2020-07-14
OpenJDK Runtime Environment (build 11.0.8+10)
```

**Comment by lochana-chathura** [comment]
- First code sample is crashing when recovering with `RESOURCE_DEF_START`.  Errorhandler keeps inserting the isolated keyword. This issue is fixed in service-typing branch and will be fixed in master once the branch is merged.

https://github.com/ballerina-platform/ballerina-lang/blob/d3b881008be4b9c9a9a21adc448c2039d52c9b30/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java#L299-L301

- Second sample was crashing due to #25724 and is fixed in current master.

**Comment by lochana-chathura** [comment]
Fixed in #27305 & #26996
